src/coreclr/vm/jitinterface.cpp

```diff
@@ -9183,10 +9179,37 @@ CorInfoType CEEInfo::getFieldTypeInternal (CORINFO_FIELD_HANDLE fieldHnd,
 
         SigPointer ptr(sig, sigCount);
 
+        // Actual field's owner
+        MethodTable* actualFieldsOwner = field->GetApproxEnclosingMethodTable();
+
+        // Potentially, a more specific field's owner (hint)
+        TypeHandle hintedFieldOwner = TypeHandle(fieldOwnerHint);
+
+        // Validate the hint:
+        bool isHintValid = false;
+        if (!hintedFieldOwner.IsNull() && !hintedFieldOwner.IsTypeDesc())
+        {
+            MethodTable* hintedFieldOwnerParent = hintedFieldOwner.AsMethodTable();
+            if (hintedFieldOwnerParent->HasSameTypeDefAs(actualFieldsOwner))
+            {
+                // if actualFieldsOwner has the same definition as hintedFieldOwner, we
+                // take hintedFieldOwner only if actualFieldsOwner is shared and hintedFieldOwner
+                // is not, hence, it's definitely more exact.
+                isHintValid = TypeHandle(actualFieldsOwner).IsCanonicalSubtype() &&
+                    !hintedFieldOwner.IsCanonicalSubtype();
+            }
+            else if (hintedFieldOwnerParent->GetMethodTableMatchingParentClass(actualFieldsOwner) != nullptr)
+            {
+                // Otherwise, use hintedFieldOwner only if it's a subclass of actualFieldsOwner
+                isHintValid = true;
+            }
+        }
+        hintedFieldOwner = isHintValid ? hintedFieldOwner : TypeHandle();
```

Comment:
```suggestion
        TypeHandle moreExactFieldOwner = TypeHandle();
        if (!hintedFieldOwner.IsNull() && !hintedFieldOwner.IsTypeDesc())
        {
            MethodTable* matchingHintedFieldOwner = hintedFieldOwner.AsMethodTable()->GetMethodTableMatchingParentClass(actualFieldsOwner);
            if (matchingHintedFieldOwner != NULL)
            {
                // we take matchingHintedFieldOwner only if actualFieldsOwner is shared and matchingHintedFieldOwner
                // is not, hence, it's definitely more exact.
                if (actualFieldsOwner->IsSharedByGenericInstantiations() &&
                    !matchingHintedFieldOwner->IsSharedByGenericInstantiations())
                {
                    moreExactFieldOwner  = matchingHintedFieldOwner;
                }
            }
        }
```
Should this be more like this? I do not think it makes sense to make a distinction between matching type and matching parent type. We should take either of those cases only if the hint is more exact.